// igniteui-i18n-core/formatters/number.formatter.js
import { BaseFormatter } from './base.formatter.js';
export class NumberFormatter extends BaseFormatter {
    constructor(defaultLocale) {
        super(defaultLocale, Intl.NumberFormat);
        this.defaultOptions = {
            maximumFractionDigits: 3,
        };
    }
    /**
     * Format a number using Intl.
     * @param value Value to be formatted.
     * @param locale Override of the current global locale.
     * @param options Options by which to format the number.
     * @returns Formatted value.
     */
    formatNumber(value, locale, options) {
        const formatter = this.getIntlFormatter(locale, options);
        return formatter.format(value);
    }
    /**
     * Get the currency symbol for provided currency code.
     * @param currencyCode The currency code to get the symbol of.
     * @param currencyDisplay How should the currency code be rendered.
     * @param locale Override locale instead of the current one.
     * @returns String representation of the currency symbol.
     */
    getCurrencySymbol(currencyCode, locale, currencyDisplay) {
        const options = {
            style: 'currency',
            currency: currencyCode,
            currencyDisplay: currencyDisplay,
            maximumFractionDigits: 0,
        };
        const formatter = this.getIntlFormatter(locale, options);
        return formatter.formatToParts(0).find((part) => part.type === 'currency')?.value;
    }
    /**
     * Get the currency symbol/name position in formatted value.
     * @param locale Override locale instead of the current one.
     * @returns Position of the symbol for number 0.
     */
    getCurrencyPosition(locale) {
        const options = {
            style: 'currency',
            currency: 'USD',
            maximumFractionDigits: 0,
        };
        const formatter = this.getIntlFormatter(locale, options);
        return formatter.formatToParts(0).findIndex((part) => part.type === 'currency');
    }
}
