// igniteui-i18n-core/i18n-manager.js
import { DateFormatter } from './formatters/date.formatter.js';
import { DisplayNamesFormatter } from './formatters/display-names.formatter.js';
import { LocaleFormatter } from './formatters/locale.formatter.js';
import { NumberFormatter } from './formatters/number.formatter.js';
import { Formatter, I18nManagerEventTarget, } from './i18n-manager.interfaces.js';
import { isBrowser } from './utils.js';
export class I18nManager extends I18nManagerEventTarget {
    _defaultLang = 'en';
    defaultLocale = 'en-US';
    static _instance;
    _currentLocale = 'en-US';
    _formatters = new Map();
    _resourcesMap = new Map([
        [
            this._defaultLang,
            {
                default: 'US',
                scripts: new Map(),
                regions: new Map([['US', {}]]),
            },
        ],
    ]);
    _rootObserver;
    get currentLocale() {
        return this._currentLocale;
    }
    get localeFormatter() {
        return this._formatters.get(Formatter.Locale);
    }
    get dateFormatter() {
        return this._formatters.get(Formatter.Date);
    }
    get numberFormatter() {
        return this._formatters.get(Formatter.Number);
    }
    get displayNamesFormatter() {
        return this._formatters.get(Formatter.DisplayNames);
    }
    static get instance() {
        if (!I18nManager._instance) {
            I18nManager._instance = new I18nManager();
        }
        return I18nManager._instance;
    }
    constructor() {
        super();
        this._formatters.set(Formatter.Locale, new LocaleFormatter(this.defaultLocale));
        this._formatters.set(Formatter.Date, new DateFormatter(this.defaultLocale, this.localeFormatter));
        this._formatters.set(Formatter.Number, new NumberFormatter(this.defaultLocale));
        this._formatters.set(Formatter.DisplayNames, new DisplayNamesFormatter(this.defaultLocale, this.dateFormatter));
        if (isBrowser()) {
            const initialLocale = document.documentElement.getAttribute('lang') ?? this.defaultLocale;
            this.setCurrentI18n(initialLocale);
            this._rootObserver = new MutationObserver((mutations) => this.htmlElementObserve(mutations));
            this._rootObserver.observe(document.documentElement, {
                attributeFilter: ['lang'],
            });
        }
    }
    /**
     * Sets up a function that will be called whenever the specified event is delivered to the target.
     * Currently supported events: onResourceChange
     * Note: Make sure to bind to events only in browser environment or emulated one. Anything else like server environment does not need to have events.
     * @param type Name of the event
     * @param listener The function to be called when event is triggered.
     * @param options
     */
    addEventListener(type, listener, options) {
        if (isBrowser()) {
            super.addEventListener(type, listener, options);
        }
    }
    /**
     * Register resource for a locale. Can be the current locale as well or a new one.
     */
    registerI18n(resources, locale) {
        const localeObj = this.getLocaleObject(locale);
        const currentResources = this.getResourcesPerLocale(locale);
        let bResourcesChanged = true;
        if (currentResources) {
            bResourcesChanged = Object.keys(resources).some((key) => resources[key] !== currentResources[key]);
            const mergedResources = { ...currentResources, ...resources };
            this.setResourcesPerLocale(locale, mergedResources);
            const defaultLocaleObj = this.localeFormatter.getIntlFormatter(this.defaultLocale);
            if (bResourcesChanged && LocaleFormatter.equalLocaleLanguages(defaultLocaleObj, localeObj)) {
                // Default locale registering new resources. Update default unfilled resources for all available languages.
                this.updateDefaultResources(mergedResources);
            }
        }
        else {
            // Fill out empty resources with available default language on register, so we don't have to fill them every time they are retrieved.
            const defaultResources = this.getDefaultResources();
            const completeResources = { ...defaultResources, ...resources };
            this.setResourcesPerLocale(locale, completeResources);
        }
        const currentLocaleObj = this.localeFormatter.getIntlFormatter(this.currentLocale);
        if (bResourcesChanged && LocaleFormatter.equalLocaleLanguages(currentLocaleObj, localeObj)) {
            this.triggerResourceChange(this.currentLocale, this.currentLocale);
        }
    }
    /**
     * Set current locale across all components.
     */
    setCurrentI18n(locale) {
        let newLocale = this.currentLocale;
        try {
            newLocale = Intl.getCanonicalLocales(locale)[0];
        }
        catch {
            console.warn(`Trying to switch to invalid locale tag '${locale}' for the Ignite UI components. Using last valid tag '${this.currentLocale}'.`);
        }
        if (this.currentLocale !== newLocale) {
            const oldLocale = this.currentLocale;
            this._currentLocale = newLocale;
            this.triggerResourceChange(oldLocale, newLocale);
            // Update formatters with latest locale.
            for (const [_, formatter] of this._formatters) {
                formatter.onLocaleChange(newLocale);
            }
        }
    }
    /**
     * Get the current resource string for all components in a single object.
     */
    getCurrentResourceStrings(locale) {
        const currentResources = this.getResourcesPerLocale(locale ?? this.currentLocale);
        if (currentResources) {
            return currentResources;
        }
        return this.getDefaultResources();
    }
    getLocaleObject(locale) {
        try {
            return this.localeFormatter.getIntlFormatter(locale);
        }
        catch {
            console.warn(`Invalid locale tag '${locale}' using for resources for the Ignite UI components. Using the default 'en-US'.`);
        }
        return this.localeFormatter.getIntlFormatter(this.defaultLocale);
    }
    getDefaultResources() {
        return this.getDefaultForCategory(this._resourcesMap.get(this._defaultLang)) ?? {};
    }
    getDefaultForCategory(category) {
        if (typeof category.default === 'string') {
            return category.scripts?.get(category.default) ?? category.regions?.get(category.default) ?? {};
        }
        return category.default;
    }
    getResourcesPerLocale(locale) {
        const localeObj = this.getLocaleObject(locale);
        const localeCategory = this._resourcesMap.get(localeObj.language);
        if (!localeCategory) {
            return undefined;
        }
        let resultResource;
        if (localeObj.script) {
            resultResource = localeCategory.scripts?.get(localeObj.script);
        }
        if (!resultResource && localeObj.region) {
            resultResource = localeCategory.regions?.get(localeObj.region);
        }
        return resultResource ?? this.getDefaultForCategory(localeCategory);
    }
    setResourcesPerLocale(locale, resources) {
        const localeObj = this.getLocaleObject(locale);
        const localeCategory = this._resourcesMap.get(localeObj.language);
        if (localeCategory) {
            if (localeObj.script) {
                localeCategory.scripts.set(localeObj.script, resources);
            }
            else if (localeObj.region) {
                localeCategory.regions.set(localeObj.region, resources);
            }
            else {
                localeCategory.default = resources;
            }
        }
        else {
            const newCategory = {
                default: {},
                scripts: new Map(),
                regions: new Map(),
            };
            if (localeObj.script) {
                newCategory.default = localeObj.script;
                newCategory.scripts.set(localeObj.script, resources);
            }
            else if (localeObj.region) {
                newCategory.default = localeObj.region;
                newCategory.regions.set(localeObj.region, resources);
            }
            else {
                newCategory.default = resources;
            }
            this._resourcesMap.set(localeObj.language, newCategory);
        }
    }
    updateDefaultResources(newDefaultResources) {
        this._resourcesMap.forEach((value, key) => {
            if (key !== this._defaultLang) {
                value.default =
                    typeof value.default === 'string' ? value.default : { ...newDefaultResources, ...value.default };
                value.scripts.forEach((scriptValue, scriptKey) => {
                    value.scripts.set(scriptKey, { ...newDefaultResources, ...scriptValue });
                });
                value.regions.forEach((scriptValue, scriptKey) => {
                    value.regions.set(scriptKey, { ...newDefaultResources, ...scriptValue });
                });
            }
        });
    }
    triggerResourceChange(oldLocale, newLocale) {
        const eventArgs = {
            oldLocale,
            newLocale,
        };
        if (isBrowser()) {
            this.dispatchEvent(new CustomEvent('onResourceChange', { detail: eventArgs }));
        }
    }
    htmlElementObserve(mutations) {
        const mutation = mutations[0];
        if (mutation.attributeName === 'lang') {
            const newLocale = mutation.target.getAttribute('lang') ?? this.currentLocale;
            this.setCurrentI18n(newLocale);
        }
    }
}
/**
 * Gets in the i18nManager instance.
 * @internal
 */
export function getI18nManager() {
    return I18nManager.instance;
}
export function getDateFormatter() {
    return getI18nManager().dateFormatter;
}
export function getNumberFormatter() {
    return getI18nManager().numberFormatter;
}
export function getDisplayNamesFormatter() {
    return getI18nManager().displayNamesFormatter;
}
/**
 * Register resources for a specific locale.
 * @param resourceStrings Object containing the translated resource strings.
 * @param locale The name of a locale. A string using the BCP 47 language tag. If not a valid one, will register to the default 'en-US' locale.
 */
export function registerI18n(resourceStrings, locale) {
    getI18nManager().registerI18n(resourceStrings, locale);
}
/**
 * Set the current locale of all Ignite UI components.
 * @param locale The name of the locale. A string using the BCP 47 language tag.
 */
export function setCurrentI18n(locale) {
    getI18nManager().setCurrentI18n(locale);
}
export function getCurrentI18n() {
    return getI18nManager().currentLocale;
}
export function getCurrentResourceStrings() {
    return getI18nManager().getCurrentResourceStrings();
}
